sos/report/plugins/mellanox_firmware.py

```diff
@@ -0,0 +1,126 @@
+
+        for device in devices:
+            self.add_cmd_output("mlxdump -d %s pcie_uc --all" % device,
+                                pred=mft_pred)
+            self.add_cmd_output("mstconfig -d %s -e q" % device,
+                                pred=mft_pred)
+            self.add_cmd_output("flint -d %s dc" % device,
+                                pred=mft_pred)
+            self.add_cmd_output("flint -d %s q" % device,
+                                pred=mft_pred)
+            self.add_cmd_output("mlxreg -d %s --reg_name "
+                                "ROCE_ACCL --get" % device,
+                                pred=mft_mlxfw_pred)
+            self.add_cmd_output("mlxlink -d %s" % device,
+                                pred=mft_mlxfw_pred)
+            self.add_cmd_output("fwtrace -d %s -i all --tracer_mode FIFO"
+                                % device, pred=mft_pred, timeout=30)
+
+            self.add_cmd_output("mstconfig -d %s -e q" % device,
+                                pred=mstflint_pred)
+            self.add_cmd_output("mstflint -d %s dc" % device,
+                                pred=mstflint_pred)
+            self.add_cmd_output("mstflint -d %s q" % device,
+                                pred=mstflint_pred)
+            self.add_cmd_output("mstreg -d %s --reg_name "
+                                "ROCE_ACCL --get" % device,
+                                pred=mstflint_mlxfw_pred)
+            self.add_cmd_output("mstlink -d %s" % device,
+                                pred=mstflint_mlxfw_pred)
+
+            # Dump the output of the mstdump command three times
+            # waiting for one second. This output is useful to check
+            # if certain registers changed
+            for i in range(3):
+                self.add_cmd_output("mstdump %s" % device, pred=mft_pred)
+                time.sleep(1)
+
+        if self.get_option('allow_system_changes'):
+            co = self.collect_cmd_output('flint --version')
+            if co['status'] != 0:
+                return
+            co = self.collect_cmd_output('mst start')
+            if co['status'] != 0:
+                return
+            co = self.collect_cmd_output('mst cable add')
+            if co['status'] != 0:
+                return
+            self.add_cmd_output("mst status -v", timeout=10)
+            self.add_cmd_output("mlxcables", timeout=10)
+            co = os.listdir("/dev/mst")
+            mlxcables = []
+            for device in co:
+                if 'cable' in device:
+                    mlxcables.append(device)
+            for mlxcable in mlxcables:
+                self.add_cmd_output("mlxcables -d %s --DDM" % mlxcable,
+                                    timeout=10)
+                self.add_cmd_output("mlxcables -d %s --dump" % mlxcable,
+                                    timeout=10)
+            self.add_cmd_output("mst stop", changes=True)
```

Comment:
I'd need some more information/context on what is happening here. `--allow-system-changes` is generally meant to gate kernel mods being loaded or the like. We don't want sos making configuration changes even if this option is used, as that is out of scope for the intended use of both sos and the option.

Reply:
In my previous role, I was using Mellanox IB regularly, and one of the things we had to do to grab extra debug data was to start `mst` to grab the extra debugs, and then stop it. This service is not started by default, and used for debugging from what I can remember. For me, this is a +1

Reply:
To add a bit more context, maybe I should add it as a comment, running `mst start` will load two kernel modules:
`mst_pci`
`mst_pciconf`
which will allow the capture, as @arif-ali described above, additional debug information.
There are some checks in order to validate the pci configuration after the two kernel modules are started.
running `mst stop` will unload those kernel modules.

Reply:
Hmm, ok. I'm onboard with that, however I think this is a better use case for the (relatively new) `collect()` approach instead of `setup()`.

From a technical perspective, when you do an `add_cmd_output()` within `setup()`, you're adding those commands to a list that gets executed later in our run after all plugins have finished their setup. Once those collections are done we move into a "manual collections" phase which invokes each plugin's `collect()` method if defined. Usually this is used to "manually" generate or compile data and then write that to a plugin-generated file via `Plugin.collection_file`, however I'm thinking that we'd prefer to have a service start/stop controlled within this phase. @pmoravec feel free to chime in if you disagree and think we should keep this in `setup()`.

Practically, what that means here is that instead of doing collections with `add_cmd_output()`, you would use `collect_cmd_output()` to execute it right then and there and save it to the archive, while using `exec_cmd()` for anything that needs to be run but you don't care about saving the output. E.G. perhaps something like this:

```python
def collect(self):
    if not self.get_option('allow_system_changes'):
        return
    self.exec_cmd('mst start')
    self.collect_cmd_output('mst status -v')
    [... etc you get the idea ...]
    self.exec_cmd('mst stop')
```